includes/lib/class-faster-image-b52f1a8-faster-image.php

```diff
@@ -0,0 +1,189 @@
+
+    /**
+     * Get the size of each of the urls in a list
+     *
+     * @param array $urls
+     *
+     * @return array
+     * @throws \Exception
+     */
+    public function batch(array $urls)
+    {
+
+        $multi   = curl_multi_init();
+        $results = array();
+
+        // Create the curl handles and add them to the multi_request
+        foreach ( array_values($urls) as $count => $uri ) {
+
+            $results[$uri] = array();
+
+            $$count = $this->handle($uri, $results[$uri]);
+
+            $code = curl_multi_add_handle($multi, $$count);
+
+            if ( $code != CURLM_OK ) {
+                throw new \Exception("Curl handle for $uri could not be added");
+            }
+        }
+
+        // Perform the requests
+        do {
+            while ( ($mrc = curl_multi_exec($multi, $active)) == CURLM_CALL_MULTI_PERFORM ) ;
+            if ( $mrc != CURLM_OK && $mrc != CURLM_CALL_MULTI_PERFORM ) {
+                throw new \Exception("Curl error code: $mrc");
+            }
+
+            if ( $active && curl_multi_select($multi) === -1 ) {
+                // Perform a usleep if a select returns -1.
+                // See: https://bugs.php.net/bug.php?id=61141
+                usleep(250);
+            }
+        } while ( $active );
+
+        // Figure out why individual requests may have failed
+        foreach ( array_values($urls) as $count => $uri ) {
+            $error = curl_error($$count);
+
+            if ( $error ) {
+                $results[$uri]['failure_reason'] = $error;
+            }
+        }
+
+        return $results;
+    }
+
+    /**
+     * @param $seconds
+     */
+    public function setTimeout($seconds)
+    {
+        $this->timeout = $seconds;
+    }
+
+    /**
+     * Create the handle for the curl request
+     *
+     * @param $url
+     * @param $result
+     *
+     * @return resource
+     */
+    protected function handle($url, & $result)
+    {
+        $stream           = new Stream_17b32f3_Stream();
+        $parser           = new Faster_Image_B52f1a8_Image_Parser($stream);
+        $result['rounds'] = 0;
+        $result['bytes']  = 0;
+        $result['size']   = 'failed';
+
+        $ch = curl_init();
+        curl_setopt($ch, CURLOPT_URL, $url);
+        curl_setopt($ch, CURLOPT_HEADER, 0);
+        curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
+        curl_setopt($ch, CURLOPT_BUFFERSIZE, 256);
+        curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, 0);
+        curl_setopt($ch, CURLOPT_SSL_VERIFYHOST, 0);
+        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
+        curl_setopt($ch, CURLOPT_CONNECTTIMEOUT, $this->timeout);
+        curl_setopt($ch, CURLOPT_TIMEOUT, $this->timeout);
+
+        #  Some web servers require the useragent to be not a bot. So we are liars.
+        curl_setopt($ch, CURLOPT_USERAGENT, 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/27.0.1453.110 Safari/537.36');
```

Comment:
I think we should probably either use the same user agent that WordPress' HTTP library does (see WordPress HTTP API) or maybe something specific to the plugin (like `amp-wp, v0.3`)?